src/rebar_pkg_resource.erl

```diff
@@ -188,37 +176,38 @@ ssl_opts(Url) ->
 %%------------------------------------------------------------------------------
 -spec store_etag_in_cache(File, ETag) -> Res when
       File :: file:name(),
-      ETag :: string(),
+      ETag :: binary(),
       Res :: ok.
 store_etag_in_cache(Path, ETag) ->
     _ = file:write_file(Path, ETag).
 
 %%%=============================================================================
 %%% Private functions
 %%%=============================================================================
--spec cached_download(TmpDir, CachePath, Pkg, Url, ETag, State, ETagPath,
+-spec cached_download(TmpDir, CachePath, Pkg, ETag, State, ETagPath,
                       UpdateETag) -> Res when
       TmpDir :: file:name(),
       CachePath :: file:name(),
       Pkg :: {pkg, Name :: binary(), Vsn :: binary(), Hash :: binary()},
-      Url :: string(),
-      ETag :: false | string(),
+      ETag :: binary(),
       State :: rebar_state:t(),
       ETagPath :: file:name(),
       UpdateETag :: boolean(),
       Res :: download_result().
-cached_download(TmpDir, CachePath, Pkg={pkg, Name, Vsn, _Hash}, Url, ETag,
+cached_download(TmpDir, CachePath, Pkg={pkg, Name, Vsn, _Hash}, ETag,
                 State, ETagPath, UpdateETag) ->
-    case request(Url, ETag) of
+    Resources = rebar_state:resources(State),
+    #{hex_config := HexConfig} = rebar_resource:find_resource_state(pkg, Resources),
+    case request(HexConfig, Name, Vsn, ETag) of
         {ok, cached} ->
             ?INFO("Version cached at ~ts is up to date, reusing it", [CachePath]),
             serve_from_cache(TmpDir, CachePath, Pkg, State);
         {ok, Body, NewETag} ->
             ?INFO("Downloaded package, caching at ~ts", [CachePath]),
             maybe_store_etag_in_cache(UpdateETag, ETagPath, NewETag),
-            serve_from_download(TmpDir, CachePath, Pkg, NewETag, Body, State,
-                                ETagPath);
-        error when ETag =/= false ->
+            serve_from_download(TmpDir, CachePath, Pkg, NewETag, Body, 
+                                State, ETagPath);
+        error when ETag =/= <<>> ->
```

Comment:
is it really nicer to have an implicit empty etag value (`<<>>`) rather than a dedicated type for invalid values?

Reply:
Not sure, I'll look at it again. Pretty sure I realized it could never be `false` with the current code so changed it there.

Reply:
No, I think this is right. `etag/1` has to return something that can be used in the request (seems the easier way to go instead of handling a `false` when making the request).

The idea here seems to be that if we have a cached copy but can't verify the etag because of some network error we still use the cached copy.

Reply:
Yeah I'm still not sure if it's great to go with binary() where <<>> is a special case or just binary() | undefined